SYCL/ESIMD/api/functional/ctors/ctor_array.cpp

```diff
@@ -0,0 +1,188 @@
+    simd_by_const_ref.copy_to(out);
+  }
+};
+
+// The main test routine.
+// Using functor class to be able to iterate over the pre-defined data types.
+template <typename DataT, int NumElems, typename TestCaseT> class run_test {
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+
+    bool passed = true;
+    const std::vector<DataT> ref_data = generate_ref_data<DataT, NumElems>();
+
+    // If current number of elements is equal to one, then run test with each
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
+      for (size_t i = 0; i < ref_data.size(); ++i) {
+        passed &= run_verification(queue, {ref_data[i]}, data_type);
+      }
+    } else {
+      passed &= run_verification(queue, ref_data, data_type);
+    }
+    return passed;
+  }
+
+private:
+  bool run_verification(sycl::queue &queue, const std::vector<DataT> &ref_data,
+                        const std::string &data_type) {
+    assert(ref_data.size() == NumElems &&
+           "Reference data size is not equal to the simd vector length.");
+
+    bool passed = true;
+
+    shared_allocator<DataT> allocator(queue);
+    shared_vector<DataT> result(NumElems, allocator);
+    shared_vector<DataT> shared_ref_data(ref_data.begin(), ref_data.end(),
+                                         allocator);
+
+    queue.submit([&](sycl::handler &cgh) {
+      const DataT *const ref = shared_ref_data.data();
+      DataT *const out = result.data();
+
+      cgh.single_task<ctors::Kernel<DataT, NumElems, TestCaseT>>(
+          [=]() SYCL_ESIMD_KERNEL {
+            DataT ref_on_dev[NumElems];
+            for (size_t i = 0; i < NumElems; ++i) {
+              ref_on_dev[i] = ref[i];
+            }
+
+            TestCaseT::template call_simd_ctor<DataT, NumElems>(
+                std::move(ref_on_dev), out);
+          });
+    });
+    queue.wait_and_throw();
+
+    for (size_t i = 0; i < result.size(); ++i) {
+      if (!are_bitwise_equal(ref_data[i], result[i])) {
+        passed = false;
+
+        const auto description =
+            ctors::TestDescription<DataT, NumElems, TestCaseT>(
+                i, result[i], ref_data[i], data_type);
+        log::fail(description);
+      }
+    }
+
+    return passed;
+  }
+};
+
+int main(int, char **) {
+  sycl::queue queue(esimd_test::ESIMDSelector{},
+                    esimd_test::createExceptionHandler());
+
+  bool passed = true;
+
+  const auto types = get_tested_types<tested_types::all>();
+  const auto dims = get_all_dimensions();
+
+  passed &= for_all_types_and_dims<run_test, initializer>(types, dims, queue);
+  passed &= for_all_types_and_dims<run_test, var_decl>(types, dims, queue);
+  passed &= for_all_types_and_dims<run_test, rval_in_expr>(types, dims, queue);
+  passed &= for_all_types_and_dims<run_test, const_ref>(types, dims, queue);
```

Comment:
```suggestion
  // Run for all combinations possible
  passed &= for_all_types_and_dims<run_test, initializer>(types, dims, queue);
  passed &= for_all_types_and_dims<run_test, var_decl>(types, dims, queue);
  passed &= for_all_types_and_dims<run_test, rval_in_expr>(types, dims, queue);
  passed &= for_all_types_and_dims<run_test, const_ref>(types, dims, queue);
```
Let's make the cases of all combinations used explicit

Reply:
I've done it in this commit: 1c8437a80d85a14ddc0ec72719271031557b3dda